blogposts/2021/integration-testing.md

```diff
@@ -0,0 +1,115 @@
+---
+title: 'How we added integration testing to our CI pipeline'
+externalTitle: 'Adding integration testing to our CI'
+description: 'Here's the story, and the lessons learned, from our work to remove all existing backend-related end-to-end tests and reliably run their corresponding unit and/or integration tests as part of our CI pipeline on all branches. 
+externalDescription: 'Integration testing isn't glamorous work, but it is impactful. Here's a behind-the-scenes look at how we worked toward a more reliable, less flaky testing suite.'
+author: Joe Chen
+publishDate: 2021-12-22T18:00+02:00
+tags: [blog]
+slug: integration-testing
+heroImage: https://storage.googleapis.com/sourcegraph-assets/blog/how-not-to-break-a-search-engine-unglamorous-engineering.jpg
+published: true
+---
+
+![Unglamorous engineering graphic](https://storage.googleapis.com/sourcegraph-assets/blog/how-not-to-break-a-search-engine-unglamorous-engineering.jpg)
+
+# How we added integration testing to our CI pipeline
+
+My name is Joe Chen and I’m on the Core Application team here at Sourcegraph. It’s our responsibility to cover all the needs of the other teams. Some of our previous work includes licensing enforcement, background permissions syncing, and an explicit permissions API.
+
+In April of 2020, we decided to tackle integration testing. Integration testing isn’t something any single team could own because it involves code-based infrastructure. Therefore, it fell on the Core Application team.
+
+The goal was to remove all existing backend-related end-to-end tests and reliably run their corresponding unit and/or integration tests as part of our CI pipeline on all branches, the end result being a more reliable, less flaky testing suite.
+
+Integration isn’t glamorous work, but it’s necessary and, most importantly, impactful. Inspired by a previous post, H[ow not to break a search engine or: What I learned about unglamorous engineering](https://about.sourcegraph.com/blog/how-not-to-break-a-search-engine-unglamorous-engineering/), I wanted to share another piece of unglamorous engineering that, despite its lack of glamor, made a big impact.
+
+## Before integration testing
+
+Our work started with an [RFC](https://handbook.sourcegraph.com/company-info-and-process/communication/rfcs), or Request For Comment, that you can see [here](https://docs.google.com/document/d/1LfCDPZZAkP4gFB0no-0Hb90EqMItnzv3YEb07B7WtrM/edit#heading=h.trqab8y0kufp).
+
+Before integration testing, we were relying entirely on end-to-end tests written in Javascript. These tests would boot up a virtual browser, move the mouse, and click buttons–simulating a real user using the instance.
+
+The problem with this method is that it’s not always stable and it’s easy to either time out or miss the button. The CI pipeline ends up becoming undependable, incurring a loss of trust in the end-to-end tests. Instead, we wanted to remove the browser manipulation layer and hit the APIs directly.
```

Comment:
```suggestion
The problem with this method is that it’s not always stable and it’s easy to either timeout or miss the button. The CI pipeline ends up becoming undependable, incurring a loss of trust in the end-to-end tests. Instead, we wanted to remove the browser manipulation layer and hit the APIs directly.
```

Reply:
I think this is a grammatical issue – "timeout" is a noun, whereas here it's used as a verb so I would have it as two words.